### ISSUE TITLE
Running `npm test` with additional arguments 

### ISSUE THREAD

**Issue by BarryThePenguin** [ember-cli/ember-cli#9084]
Looking at #9009 there are a lot of great changes

- replacing `ember-cli-eslint` with plain old `eslint`
- running commands in parallel
- a nice convention of `lint:*` and `test:*` npm scripts

Great work all round 🎉

I looked at adopting this approach in our projects. We're not using travis for CI, so that's one difference.. but we do run `npm test` (technically `yarn test`) in CI, which is what the above PR encourages

> All too many times I’ve been bitten by addons that are not using the default .travis.yml setup (using something like circle, GH actions, or other node aware tooling) that assume `npm test` means “good to go”.

We follow the same idea. Running `yarn test` in CI means everything is good, including tests, linting, etc.. 👍 However, I noticed in the PR that addons use both `npm lint` and `npm test:ember` in travis

https://github.com/ember-cli/ember-cli/blob/28be1a59c4a722d3fdda4ff492e38a944ca6b479/blueprints/addon/files/.travis.yml#L40-L42

Is that how we should move forward? Doesn't that deviate from `npm test` being the signal for "good to go"?

Where we deviate from the blueprint is we're using `"test:ember": "ember exam"`. According to the [documentation](https://github.com/ember-cli/ember-exam), this it's a drop in replacement

> Using Ember Exam is fairly straightforward as it extends directly from the default Ember-CLI test command. So, by default, it will work exactly the same as ember test.

Where we run into trouble is, we pass additional cli arguments to `yarn test`, namely `yarn test --test-port=${randomTestPort} --reporter=dot --split=4 --parallel=1 --random`. Some of these are ember exam specific, but that's not an issue.

The problem I've run into is it's not clear how additional cli arguments should be passed into `npm test`. I've read the [`npm-run-all` documentation](https://github.com/mysticatea/npm-run-all/blob/HEAD/docs/npm-run-all.md#run-with-arguments), and I'm still not sure how we could achieve my example above?

**Comment by rwjblue** [comment]
> However, I noticed in the PR that addons use both npm lint and npm test:ember in travis
> 
> ```yml
>  script: 
>    - <% if (yarn) { %>yarn<% } else { %>npm run<% } %> lint 
>    - <% if (yarn) { %>yarn<% } else { %>npm run<% } %> test:ember 
> ```
> 
> Is that how we should move forward? Doesn't that deviate from npm test being the signal for "good to go"?

The stated goal of `npm test` / `yarn test` being the source of truth has not changed. If you run `npm test` or `yarn test` locally, you would indeed see the same exact failures that the provided `.travis.yml` has. The subtle (yet IMHO, important) distinction here is that CI can do something a tad bit more intelligent than we can support for running local: leverage multiple builds. The goal of the `.travis.yml` setup is to have an initial set of "does it work" smoke testing (running linting, and `ember test` with the default ember-source version) before spawning off individual builds for all of the supported ember-try scenarios.

To clarify: `npm test` / `yarn test` are the source of truth, but CI configurations may be specially curated to leverage the capabilities of the individual CI platform.

**Comment by rwjblue** [comment]
> Looking at #9009 there are a lot of great changes
> 
> * replacing `ember-cli-eslint` with plain old `eslint`
> * running commands in parallel
> * a nice convention of `lint:*` and `test:*` npm scripts
> 
> Great work all round 🎉
> 
> I looked at adopting this approach in our projects. We're not using travis for CI, so that's one difference.. but we do run `npm test` (technically `yarn test`) in CI, which is what the above PR encourages
> 
> > All too many times I’ve been bitten by addons that are not using the default .travis.yml setup (using something like circle, GH actions, or other node aware tooling) that assume `npm test` means “good to go”.
> 
> We follow the same idea. Running `yarn test` in CI means everything is good, including tests, linting, etc.. 👍 However, I noticed in the PR that addons use both `npm lint` and `npm test:ember` in travis
> 
> https://github.com/ember-cli/ember-cli/blob/28be1a59c4a722d3fdda4ff492e38a944ca6b479/blueprints/addon/files/.travis.yml#L40-L42
> 
> Is that how we should move forward? Doesn't that deviate from `npm test` being the signal for "good to go"?
> 
> Where we deviate from the blueprint is we're using `"test:ember": "ember exam"`. According to the [documentation](https://github.com/ember-cli/ember-exam), this it's a drop in replacement
> 
> > Using Ember Exam is fairly straightforward as it extends directly from the default Ember-CLI test command. So, by default, it will work exactly the same as ember test.
> 
> Where we run into trouble is, we pass additional cli arguments to `yarn test`, namely `yarn test --test-port=${randomTestPort} --reporter=dot --split=4 --parallel=1 --random`. Some of these are ember exam specific, but that's not an issue.
> 
> The problem I've run into is it's not clear how additional cli arguments should be passed into `npm test`. I've read the [`npm-run-all` documentation](https://github.com/mysticatea/npm-run-all/blob/HEAD/docs/npm-run-all.md#run-with-arguments), and I'm still not sure how we could achieve my example above?

If you _always_ want to pass those arguments to `test:ember`, then I personally would change `"test:ember": "ember exam",` into `"test:ember": "ember exam --test-port=0 --reporter=dot --split=4 --parallel=1 --random",`. If you use this modification to your `script`, then `npm test` / `yarn test` still continue to work properly locally and callers do not need to provide custom arguments to every invocation.

Are you more asking how you _could_ pass arguments to individual script calls with `npm-run-all` (as in, you don't want the default `yarn test`/`npm test` to include these arguments, but you do want to allow passing them forward)?

**Comment by BarryThePenguin** [comment]
> Are you more asking how you _could_ pass arguments to individual script calls with `npm-run-all` (as in, you don't want the default `yarn test`/`npm test` to include these arguments, but you do want to allow passing them forward)?

Yes, that summarises my question well 😅

Reading the documentation, I wasn't able to come up with a `yarn test` incantation that allowed me to pass arguments to `ember exam`

> To clarify: `npm test` / `yarn test` are the source of truth, but CI configurations may be specially curated to leverage the capabilities of the individual CI platform.

This is a great point 👍 which probably solves my issue..

1. `npm test` is the "source of truth"
1. CI can do the same thing, leveraging multiple builds

Maybe this can be made clearer through documentation somewhere?

**Comment by rwjblue** [comment]
> Maybe this can be made clearer through documentation somewhere?

Ya, I agree. We should probably make a section in the addon parts of cli.emberjs.com..

**Comment by locks** [comment]
This whole area has change many times since this issue was originally submitted. If you encounter any problems with the current versions please open new issue reports. Thanks!
